pkg/api/handlers/libpod/containers_stats.go

```diff
@@ -62,25 +62,40 @@ func StatsContainer(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	// Write header and content type.
-	w.WriteHeader(http.StatusOK)
-	w.Header().Set("Content-Type", "application/json")
-	if flusher, ok := w.(http.Flusher); ok {
-		flusher.Flush()
-	}
-
+	wroteContent := false
 	// Set up JSON encoder for streaming.
 	coder := json.NewEncoder(w)
 	coder.SetEscapeHTML(true)
 
 	for stats := range statsChan {
+		if !wroteContent {
+			if stats.Error != nil {
+				utils.ContainerNotFound(w, "", stats.Error)
+				return
+			}
+		}
 		if err := coder.Encode(stats); err != nil {
 			// Note: even when streaming, the stats goroutine will
 			// be notified (and stop) as the connection will be
 			// closed.
-			logrus.Errorf("Unable to encode stats: %v", err)
+			if !wroteContent {
+				utils.InternalServerError(w, fmt.Errorf("failed to encode stats: %w", err))
+				return
+			}
+			logrus.Errorf("Failed to encode stats: %v", err)
 			return
 		}
+
+		if !wroteContent {
+			// Write header and content type.
+			w.WriteHeader(http.StatusOK)
+			w.Header().Set("Content-Type", "application/json")
+			if flusher, ok := w.(http.Flusher); ok {
+				flusher.Flush()
+			}
+			wroteContent = true
+		}
```

Comment:
this would need happen before the coder.encode call as headers/status code must be written before content.
and you can remove the flush here as it flushes below anyway.